activerecord/lib/active_record/connection_adapters/abstract/query_comment.rb

```diff
@@ -0,0 +1,143 @@
+        if comment.present? && !sql.include?(comment)
+          sql = if prepend_comment
+            "#{comment} #{sql}"
+          else
+            "#{sql} #{comment}"
+          end
+        end
+        sql
+      end
+
+      # Maintains a user-defined context for all queries and constructs an SQL comment by
+      # calling methods listed in :components
+      #
+      # Additional information can be pushed to the context to be referenced by methods
+      # defined in framework or application initializers.
+      module Context
+
+        mattr_accessor :components, :lines_to_ignore
+        mattr_accessor :cache_query_comment, default: false
+        thread_mattr_accessor :context, default: {}
+        thread_mattr_accessor :cached_comment
+        DEFAULT_LINES_TO_IGNORE_REGEX = %r{\.rvm|/ruby/gems/|vendor/|active_record|active_support|rbenv|monitor\.rb.*mon_synchronize}
+
+        class << self
+          Context.components = [:application]
+
+          def comment
+            return cached_comment if cache_query_comment && !cached_comment.nil?
+            self.cached_comment = "/*#{escape_sql_comment(comment_content)}*/"
+          end
+
+          def escape_sql_comment(str)
+            while str.include?('/*') || str.include?('*/')
+              str = str.gsub('/*', '').gsub('*/', '')
+            end
+            str
+          end
+
+          def comment_content
+            components.filter_map do |c|
+              if component_value = self.try(c)
+                "#{c}:#{component_value}"
+              end
+            end.join(",")
+          end
+
+          def set(ctx)
+            return unless ctx.is_a? Hash
+            self.context.merge!(ctx)
+            self.cached_comment = nil
+          end
+
+          def context_available?(context_name)
+            context.key?(context_name) && context[context_name].present?
+          end
+
+          def clear_comment_cache!
+            self.cached_comment = nil
+          end
+
+          # Default ActiveRecord comment components
+          # :application, :connection_config, :line
+          def application
+            if defined?(Rails.application)
+              context[:application_name] ||= Rails.application.class.name.split("::").first
+            else
+              context[:application_name] ||= "rails"
+            end
+            context[:application_name]
+          end
+
+          def connection_config
+            return unless context_available?(:adapter) && context[:adapter].pool
+            context[:adapter].pool.db_config.configuration_hash
+          end
+
+          def socket
+            if connection_config.present?
+              connection_config[:socket]
+            end
+          end
+
+          def db_host
+            if connection_config.present?
+              connection_config[:host]
+            end
+          end
+
+          def database
+            if connection_config.present?
+              connection_config[:database]
```

Comment:
The majority of these have methods defined on `db_config` and that's how they should be accessed rather than accessing the hash (ie db_config.databas). These will also handle the defaults in the scenario these values are missing from the hash. It's probably not clear from the way db_config works, but we add reader methods for anything Rails needs to access - Rails should never access the hash directly except for when it gets passed to the client. So for socket, which doesn't have a reader method, we should add one so Rails doesn't need to access the hash directly.

Reply:
Ah good to know. I'm not even sure if the current Marginalia implementation of these helpers works with current Rails. It refers to `pool.spec.config`, so I think it's fine to rename / use `db_config` directly in the context, and then in the helper methods call the methods defined on `db_config`.

Reply:
Oh yea if it refers to `pool.spec` that doesn't work in current Rails - spec is gone.

Reply:
> we should add one so Rails doesn't need to access the hash directly.

I've added a `socket` method to `HashConfig` - given it has no defaults (like `database` & `host`) I haven't added any coverage. LMK if it should be more involved, or if I should submit that in a separate PR.

At the very least it provides a stubbing point :)